src/bundle.c

```diff
@@ -1625,6 +1625,136 @@ gboolean check_bundle_payload(RaucBundle *bundle, GError **error)
+	GError *ierror = NULL;
+	gboolean res = FALSE;
+	gssize size;
+
+	g_return_val_if_fail(bundle != NULL, FALSE);
+	g_return_val_if_fail(outpath != NULL, FALSE);
+	g_return_val_if_fail(insig != NULL, FALSE);
+	g_return_val_if_fail(error == NULL || *error == NULL, FALSE);
+
+	r_context_begin_step("replace_signature", "Replacing bundle signature", 1);
+
+	tmpdir = g_dir_make_tmp("rauc-replace-XXXXXX", &ierror);
+	if (tmpdir == NULL) {
+		g_propagate_prefixed_error(error, ierror,
+				"Failed to create tmp dir: ");
+		res = FALSE;
+		goto out;
+	}
+
+	iname = g_build_filename(tmpdir, "_bundle.raucb", NULL);
+	ifile = g_file_new_for_path(iname);
+	istream = g_file_create(ifile, G_FILE_CREATE_PRIVATE, NULL, &ierror);
+	if (istream == NULL) {
+		g_propagate_prefixed_error(
+				error,
+				ierror,
+				"failed to create the temporary bundle to replace signature: ");
+		goto out;
+	}
+
+	sig = read_file(insig, &ierror);
+	if (!sig) {
+		g_propagate_prefixed_error(
+				error,
+				ierror,
+				"failed to read signature file: ");
+		res = FALSE;
+		goto out;
+	}
+
+	res = g_seekable_seek(G_SEEKABLE(bundle->stream),
+			0, G_SEEK_SET, NULL, &ierror);
+	if (!res) {
+		g_propagate_prefixed_error(
+				error,
+				ierror,
+				"failed to seek to start of original bundle: ");
+		goto out;
+	}
+
+	size = g_output_stream_splice(
+			(GOutputStream*)istream,
+			bundle->stream,
+			G_OUTPUT_STREAM_SPLICE_NONE,
+			NULL, &ierror);
+	if (size == -1) {
+		g_propagate_prefixed_error(
+				error,
+				ierror,
+				"failed to append original bundle payload to temporary bundle: ");
+		res = FALSE;
+		goto out;
+	}
+
+	res = g_seekable_seek(G_SEEKABLE((GOutputStream*)istream),
+			bundle->size, G_SEEK_SET, NULL, &ierror);
+	if (!res) {
+		g_propagate_prefixed_error(
+				error,
+				ierror,
+				"failed to seek to payload end offset of temporary bundle: ");
+		goto out;
+	}
+
+	res = output_stream_write_bytes_all((GOutputStream*)istream, sig, NULL, &ierror);
+	if (!res) {
+		g_propagate_prefixed_error(
+				error,
+				ierror,
+				"failed to append signature to temporary bundle: ");
+		goto out;
+	}
+
+	size = (guint64)g_bytes_get_size(sig);
+	res = output_stream_write_uint64_all((GOutputStream*)istream, size, NULL, &ierror);
+	if (!res) {
+		g_propagate_prefixed_error(
+				error,
+				ierror,
+				"failed to append signature size to temporary bundle: ");
+		goto out;
+	}
+
+	/* Necessary to release associated fd before perform check_bundle */
+	g_object_unref(istream);
+	res = check_bundle(iname, &ibundle, verify, &ierror);
+	if (!res) {
+		g_propagate_prefixed_error(
+				error,
+				ierror,
+				"failed to verify the temporary bundle: ");
+		goto out;
+	}
+
+	outfile = g_file_new_for_path(outpath);
+	res = g_file_copy(ifile, outfile, G_FILE_COPY_OVERWRITE, NULL, NULL, NULL,
+			&ierror);
+	if (!res) {
+		g_propagate_prefixed_error(
+				error,
+				ierror,
+				"failed to move temporary bundle to final bundle: ");
+		goto out;
+	}
+
+	res = TRUE;
+out:
```

Comment:
You should remove the temporary file if it still exists here, to handle error cases. (see `out:` in `create_bundle()`).